lib/pbench/server/indexing_tarballs.py

```diff
@@ -425,13 +435,28 @@ def sigquit_handler(*args):
 
                         if sigquit_interrupt[0]:
                             break
+                        if sighup_interrupt[0]:
+                            status, tarballs = self.collect_tb()
+                            if status != 0:
+                                idxctx.logger.info(
+                                    "Unexpected error while recollecting tarballs"
```

Comment:
"recollecting" sounds odd here, and the normal sense of the word doesn't fit. How about "refreshing list of tarballs to index"??

Also, we continue on despite the failure... but on an error `collect_tb()` returns an empty tarball list, so you've forgotten whatever we started with.

I think you should try `status, new_tb = self.collect_tb()` instead, and assign `tarballs = new_tb` only if it succeeded.

Also, how does the iteration over `tarballs` work when you're completely replacing the list with a new list? Does Python handle this the way we want? (I don't know offhand.)

Finally, you're reporting how many we've done and how many remain based on "idx" relative to the number of elements in `tarballs` ... but `collect_tb()` won't find the datasets we've already indexed in the loop, so `tarballs` will be the number _remaining_ from the original list (unless more have been added). So, for example, if we find 10, index 5, and then get a `SIGHUP`, our `idx` is 5, and `collect_tb()` returns the 5 we haven't indexed yet... so your message will report that we've indexed 5 and there are 0 remaining (5 - 5). I think this logic needs to be re-worked.

While the "update the list" is an intriguing idea, remember that the indexer will be triggered again in a short time, so I'm not sure how useful that really is. The consistency issues would go away if you dropped that and just continued to iterate through the original list.

Eventually (and hopefully not that far off) the indexer will be demand-driven by messages that a new dataset is available.

Reply:
> "recollecting" sounds odd here, and the normal sense of the word doesn't fit. How about "refreshing list of tarballs to index"??

I will change it to this or webb's  suggestion is also a good one



> Also, we continue on despite the failure... but on an error `collect_tb()` returns an empty tarball list, so you've forgotten whatever we started with.

I thought outside for loop will take care of that by completing the iterations, breaking it here might create unwanted terminations



> I think you should try `status, new_tb = self.collect_tb()` instead, and assign `tarballs = new_tb` only if it succeeded.

yeah sure I thought reassigning directly will make more sense



> Also, how does the iteration over `tarballs` work when you're completely replacing the list with a new list? Does Python handle this the way we want? (I don't know offhand.)

when the signal for the SIGHUP handler is called then it completes processing of the current tarball and then checks whether the signal is interrupted. Unindexed tarballs are all present in the "To-Index" folder so it will contain all the new tarballs that just came in and also all the tarballs that would get processed after this iteration. So, if we re-check that folder then it would contain all the tarballs that need to be indexed else they will get duplicated. like we are processing [1, 2, 3, 4] and [1, 2] are indexed, "To-index" folder will contain [3, 4]  as well as new tarballs that came in [5, 6]. So, we can just check tarballs in the "to-index" folder and replace it with an existing list which is 'tarballs'.



> Finally, you're reporting how many we've done and how many remain based on "idx" relative to the number of elements in `tarballs` ... but `collect_tb()` won't find the datasets we've already indexed in the loop, so `tarballs` will be the number _remaining_ from the original list (unless more have been added). So, for example, if we find 10, index 5, and then get a `SIGHUP`, our `idx` is 5, and `collect_tb()` returns the 5 we haven't indexed yet... so your message will report that we've indexed 5 and there are 0 remaining (5 - 5). I think this logic needs to be re-worked.

yeah I didn't think it thoroughly will work on it

> While the "update the list" is an intriguing idea, remember that the indexer will be triggered again in a short time, so I'm not sure how useful that really is. The consistency issues would go away if you dropped that and just continued to iterate through the original list.

What would be best?

Reply:
> @dbutenhof, perhaps a hyphen would help -- "re-collecting"? :-D
> 
> As for reassigning `tarballs`, see my comment below on `pbench-index.py` line 186.

yeah sure